Update Navio2 Setup Guide (#2726)
PX4/PX4-user_guide: en/flight_controller/raspberry_pi_navio2.md

PX4 support for this flight controller is [experimental](../flight_controller/autopilot_experimental.md).
:::

This is the developer "quickstart" for Raspberry Pi 2/3/4 Navio2 autopilots.
It allows you to build PX4 and transfer to the RasPi, or build natively.

![Ra Pi Image](../../assets/hardware/hardware-rpi2.jpg)

## OS Image

> Use the [Emlid RT Raspbian image for Navio 2](https://docs.emlid.com/navio2/configuring-raspberry-pi) Raspberry Pi OS image.

Reviewer: ```suggestion
Use the preconfigured [Emlid Raspberry Pi OS image for Navio 2](https://docs.emlid.com/navio2/configuring-raspberry-pi).
```

## Setting up Access

The Raspberry Pi OS image has SSH setup already.
Username is "pi" and password is "raspberry".
We assume that the username and password remain at their defaults for the purpose of this guide.

To setup the Pi to join your local wifi, follow [this guide](https://www.raspberrypi.org/documentation/configuration/wireless/wireless-cli.md), or connect it via an ethernet cable.

> To connect to your pi via SSH, use the default username(pi) and hostname(navio). Alternatively if this doesn't work, you can find the IP address of your Pi and specify it.

Reviewer: ```suggestion
To connect to your RPi via SSH, use the default username (`pi`) and hostname (`navio`).
Alternatively (if this doesn't work), you can find the IP address of your RPi and specify it.
```

Author: I agree with one minor change, for most of the guide the Raspberry Pi is just referred to as a Pi, RPi is fine too just want to make sure it's consistent.

Reviewer: @johnBrereton Sorry, missed this comment when I merged this. But I had already checked and seen you made the changes consistently, so this remains Pi.

```sh
ssh pi@navio.local
```
or
```sh
ssh pi@<IP-ADDRESS>
```

## Expand the Filesystem

> Expand file system to take advantage of entire SD card by running:

Reviewer: ```suggestion
Expand the file system to take advantage of entire SD card by running:
```

## Setting up Access

The Raspberry Pi OS image has SSH setup already.
Username is "pi" and password is "raspberry".
We assume that the username and password remain at their defaults for the purpose of this guide.

To setup the Pi to join your local wifi, follow [this guide](https://www.raspberrypi.org/documentation/configuration/wireless/wireless-cli.md), or connect it via an ethernet cable.

To connect to your Pi via SSH, use the default username (`pi`) and hostname (`navio`).
Alternatively (if this doesn't work), you can find the IP address of your RPi and specify it.

```sh
ssh pi@navio.local
```

> or

Reviewer: ```suggestion

or

```

#dtoverlay=navio-rgb
```

> ## Changing Hostnames

Reviewer: This is not needed/advisable any more? I.e why do we no longer want to avoid conflicts with other RPis on the network?

Author: The pre-configured image now comes with a custom hostname and no longer uses the standard “raspberrypi” hostname which would create conflict.

Reviewer: Thank you for clarification.

```sh
echo "Hello" > hello.txt
scp hello.txt pi@navio.local:/home/pi/
rm hello.txt
```

This should copy over a "hello.txt" file into the home folder of your Pi.
Validate that the file was indeed copied, and you can proceed to the next step.

## Building the Code

> Follow the instructions below to build the source code on your developmenet machine (Ubuntu 18.04) and transfer the compiled program to the Pi, unfortunetealy, the option to build the code natively(on the Pi) is no longer available.

Reviewer: Does it have to be Ubuntu 18.04? Or is this just what you used? Would these work on Ubuntu 22.04. I ask because Ubuntu 18.04 is EOL.

Author: Correct, it must be 18.04 or older, I was skeptical at first too, but apparently the binary compiled by any newer version of Ubuntu is too new for the Pi, and the compiler scripts are only made for Ubuntu.  I tried on 22.04 as I already had a VM setup, but ran into some weird errors on the Pi (error: static assertion failed: atomic is not lock-free for the given type T).

Reviewer: Thanks. I'm going to call this out in a note.

### Cross-compiler Build

First install the [standard developer environment on your Ubunto development computer](../dev_setup/dev_env_linux_ubuntu.md#raspberry-pi).

Specify the IP (or hostname) of your Pi using:

```sh
export AUTOPILOT_HOST=navio.local
```

or

```sh
export AUTOPILOT_HOST=192.168.X.X
```

:::note

> The value of the environment variable should be set before the build, or `make upload` will fail to find your Pi.

Reviewer: I am used to seeing RPi or RasPi. I don't particularly care as long as the naming is consistent through the document.

```suggestion
The value of the environment variable should be set before the build, or `make upload` will fail to find your RPi.
```

```sh
export AUTOPILOT_HOST=pi_hostname.domain
export AUTOPILOT_HOST=192.168.X.X
```

:::note
The value of the environment variable should be set before the build, or `make upload` will fail to find your RPi.
The value of the environment variable should be set before the build, or `make upload` will fail to find your Pi.
:::

Build the executable file:
Build the executable file on your development machine:

```sh
cd PX4-Autopilot
make emlid_navio2 # for cross-compiler build
make emlid_navio2
```

The "px4" executable file is in the directory **build/emlid_navio2_default/**.
Make sure you can connect to your RPi over ssh, see [instructions how to access your RPi](#setting-up-access).
Make sure you can connect to your Pi over ssh, see [instructions how to access your Pi](#setting-up-access) following the instructions for armhf under Raspberry Pi.

Then upload it with:

```sh
cd PX4-Autopilot
make emlid_navio2 upload # for cross-compiler build
make emlid_navio2 upload
```

Then, connect over ssh and run it with (as root):
Then, connect over ssh and run it on the Pi (as root):

```sh
cd ~/px4
sudo ./bin/px4 -s px4.config
```

### Native Build

> A native build is one that you run directly on the Pi (the other option is to run builds on a development computer which cross-compiles for the Pi, and pushes the PX4 executable binary directly to the Pi).

Reviewer: Ah, I see your comment above. Thanks.

> A native build is one that you run directly on the Pi (the other option is to run builds on a development computer which cross-compiles for the Pi, and pushes the PX4 executable binary directly to the Pi).

Author: Correct, a native build is no longer supported and the PX4 repo no longer contains the make files to do so.

```sh
echo "Hello" > hello.txt
scp hello.txt pi@navio.local:/home/pi/
rm hello.txt
```

This should copy over a "hello.txt" file into the home folder of your Pi.
Validate that the file was indeed copied, and you can proceed to the next step.

## Building the Code

> Follow the instructions below to build the source code on your development machine (Ubuntu 18.04) and transfer the compiled program to the Pi.
> Note that earlier versions allowed code to be built natively (on the Pi), but this option is no longer available.

Reviewer: ```suggestion
:::note
PX4 binaries for Navio 2 can only be built on Ubuntu 18.04.
Ubuntu 20.04 and later do not currently work (as of September 2023). 
:::

Follow the instructions below to build the source code on your development machine and transfer the compiled program to the Pi.
Note that earlier versions allowed code to be built natively (on the Pi), but this option is no longer available.
```

```sh
echo "Hello" > hello.txt
scp hello.txt pi@navio.local:/home/pi/
rm hello.txt
```

This should copy over a "hello.txt" file into the home folder of your Pi.
Validate that the file was indeed copied, and you can proceed to the next step.

## Building the Code

> :::note

Reviewer: FWIW significant enough to mention up front.

PX4 binaries for Navio 2 can only be built on Ubuntu 18.04.
Ubuntu 20.04 and later do not currently work (as of September 2023). 
:::

Follow the instructions below to build the source code on your development machine and transfer the compiled program to the Pi.
Note that earlier versions allowed code to be built natively (on the Pi), but this option is no longer available.

### Cross-compiler Build

> First install the [standard developer environment on your Ubunto development computer](../dev_setup/dev_env_linux_ubuntu.md#raspberry-pi).

Reviewer: Typo
```suggestion
First install the [standard PX4 developer environment](../dev_setup/dev_env_linux_ubuntu.md#raspberry-pi) on your Ubuntu 18.04 development computer.
```

Specify the IP (or hostname) of your Pi using:

```sh
export AUTOPILOT_HOST=navio.local
```

or

```sh
export AUTOPILOT_HOST=192.168.X.X
```

:::note
The value of the environment variable should be set before the build, or `make upload` will fail to find your Pi.
:::

Build the executable file on your development machine:

```sh
cd PX4-Autopilot
make emlid_navio2
```

The "px4" executable file is in the directory **build/emlid_navio2_default/**.

> Make sure you can connect to your Pi over ssh, see [instructions how to access your Pi](#setting-up-access) following the instructions for armhf under Raspberry Pi.

Reviewer: ```suggestion
Make sure you can connect to your Pi over SSH, see [instructions how to access your Pi](#setting-up-access) following the instructions for armhf under Raspberry Pi.
```
